pkg/plugin/route_table_canary.go

```diff
@@ -0,0 +1,123 @@
+)
+
+func (r *RpcPlugin) handleCanaryUsingRouteTables(
+	ctx context.Context,
+	rollout *v1alpha1.Rollout,
+	desiredWeight int32,
+	additionalDestinations []v1alpha1.WeightDestination,
+	pluginConfig *GlooEdgeTrafficRouting) error {
+
+	rts, err := r.getRouteTables(ctx, rollout, pluginConfig)
+	if err != nil {
+		return err
+	}
+
+	allRouteTablesForCanary, err := r.getDestinationsInRouteTables(rollout, pluginConfig, rts)
+	if err != nil {
+		return err
+	}
+
+	r.maybeConvertSingleToMulti(allRouteTablesForCanary)
+	r.maybeCreateCanaryDestinations(allRouteTablesForCanary, getCanaryServiceName(rollout))
+
+	for _, rt := range allRouteTablesForCanary {
+		originalRt := &gwv1.RouteTable{}
+		rt.RouteTable.DeepCopyInto(originalRt)
+
+		for _, dst := range rt.Destinations {
+			dst.Stable.Weight = &wrapperspb.UInt32Value{Value: uint32(100 - desiredWeight)}
+			dst.Canary.Weight = &wrapperspb.UInt32Value{Value: uint32(desiredWeight)}
+		}
+
+		if err = r.Client.RouteTables().PatchRouteTable(ctx, rt.RouteTable, client.MergeFrom(originalRt)); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *RpcPlugin) getDestinationsInRouteTables(
+	rollout *v1alpha1.Rollout,
+	pluginConfig *GlooEdgeTrafficRouting,
+	routeTables []*gwv1.RouteTable) (ret []routeTableWithDestinations, err error) {
+
+	for _, rt := range routeTables {
+		if rt.Spec.GetRoutes() == nil {
+			continue
+		}
+
+		if len(rt.Spec.GetRoutes()) > 1 && len(pluginConfig.Routes) == 0 {
+			return nil,
+				fmt.Errorf("route table %s/%s has multiple routes but canary config doesn't specify which routes to use", rt.GetNamespace(), rt.GetName())
+		}
+
+		dsts := r.getDestinationsInRoutes(rt.Spec.GetRoutes(), rollout, pluginConfig)
+		if len(dsts) == 0 {
+			continue
+		}
+
+		ret = append(ret, routeTableWithDestinations{RouteTable: rt, Destinations: dsts})
+	}
+
+	if len(ret) == 0 {
+		return nil, fmt.Errorf("couldn't find stable services in RouteTables selected with Name: '%s', Namespace: '%s', Labels: %v, with route names in %v",
+			pluginConfig.RouteTableSelector.Name, pluginConfig.RouteTableSelector.Namespace, pluginConfig.RouteTableSelector.Labels, pluginConfig.Routes)
+	}
+
+	return ret, nil
+}
+
+func (r *RpcPlugin) getRouteTables(ctx context.Context, rollout *v1alpha1.Rollout, pluginConfig *GlooEdgeTrafficRouting) ([]*gwv1.RouteTable, error) {
+	namespace := pluginConfig.RouteTableSelector.Namespace
+
+	if namespace == "" {
+		r.LogCtx.Debugf("defaulting RouteTable selector namespace to Rollout namespace %s for rollout %s", rollout.Namespace, rollout.Name)
+		namespace = rollout.Namespace
+	}
+
+	if pluginConfig.RouteTableSelector.Name == "" && len(pluginConfig.RouteTableSelector.Labels) == 0 {
+		return nil, fmt.Errorf("name or labels field must be set in RouteTable selector")
+	}
+
+	if pluginConfig.RouteTableSelector.Name != "" {
+		return r.getRouteTable(ctx, namespace, pluginConfig.RouteTableSelector.Name)
+	}
+
+	return r.listRouteTables(ctx, namespace, pluginConfig)
```

Comment:
Why do we support label selections for RouteTables but not for VirtualServices?

Reply:
the gist is we support a single VS but multiple RTs; This is an MVP, figured that it's sufficient to support just multiple RTs. Multiple VS support can be added if/when customers ask for it.

Reply:
agree w/ @dmitri-d , I can see possible use cases but we can wait till customer asks